gradio/flagging.py

```diff
@@ -575,13 +576,29 @@ def dump_json(self, thing: dict, file_path: str | Path) -> None:
 
 class FlagMethod:
     """
-    Helper class that contains the flagging button option and callback
+    Helper class that contains the flagging options and calls the flagging method. Also
+    provides visual feedback to the user when flag is clicked.
     """
 
-    def __init__(self, flagging_callback: FlaggingCallback, flag_option=None):
+    def __init__(
+        self,
+        flagging_callback: FlaggingCallback,
+        label: str,
+        value: str,
+        visual_feedback: bool = True,
+    ):
         self.flagging_callback = flagging_callback
-        self.flag_option = flag_option
+        self.label = label
+        self.value = value
         self.__name__ = "Flag"
+        self.visual_feedback = visual_feedback
 
     def __call__(self, *flag_data):
-        self.flagging_callback.flag(list(flag_data), flag_option=self.flag_option)
+        try:
+            self.flagging_callback.flag(list(flag_data), flag_option=self.value)
+        except Exception as e:
+            raise gr.Error("Error while flagging: {}".format(e))
+        if not self.visual_feedback:
+            return
+        time.sleep(0.8)  # to provide enough time for the user to observe button change
+        return gr.Button.update(value=self.label, interactive=True)
```

Comment:
The only problem I see here is that if there is an exception, the button won't change back.

I really like the idea of raising gr.Error here though. So I see two solutions:

1. Be able to `return` a gr.Error instead of just raising it to give visual feedback about the error and return the other updates along side it.
2. Add a  `Press Clear to reset Flag button message to the exception` and modify the Clear button to reset the sates of the flag buttons.

What do you think?

Reply:
Ah good catch @freddyaboulton. (1) would be cool but I think it would make return values from functions more complicated. I suppose the way to do it would be that you can return a `gr.Error` along with the other return values and it is skipped over when assigning return values to output components. We could consider that but it's definitely outside the scope of this PR. I'll implement (2)

Reply:
So this is actually  trickier than I thought. The reason is that the `Button` component does not have a `StatusTracker`, so it cannot actually show the `gr.Error` modal unlike the other components. I tried a few different things:

(1) adding a "hidden" output component which could be the target of the `gr.Error` --> however, hidden components don't show the `StatusTracker` either
(2) making Interface's output components the target of the `gr.Error` --> however, this means that the output components show a loading indicator when the flag button is clicked, which is not ideal
(3) adding a `StatusTracker` to the `Button` component --> however, this messed up the layout of the `Button` (doable but would require a bit of frontend work)

In the end, I've just removed the `gr.Error` for now --> instead, if an error is raised, the `Button` just displays the string "Error!", the error is printed to the console / logs, and you can press the clear button to reset flagging. Given that errors during flagging should be pretty rare, I think this should be good enough for now

Reply:
To test, run a simple Interface like this:

```py
import gradio as gr

gr.Interface(lambda x:x, "textbox", "textbox", flagging_options=["a", "b"]).launch()
```

After the interface launches, delete the "flagged" directory that has been created in the same directory as the Gradio app script. Then, enter something into the input textbox and try flagging it. You'll see the error message appear on the button like this:

<img width="580" alt="image" src="https://user-images.githubusercontent.com/1778297/221760572-81a7ae0f-e018-42a1-aff5-32e8d6342b31.png">

Reply:
> So this is actually trickier than I thought. The reason is that the Button component does not have a StatusTracker, so it cannot actually show the gr.Error modal unlike the other components.

I actually think this is a design error, the global error model should not require a status tracker. Will file an issue.

Reply:
#3347

Reply:
Thanks @pngwn! Happy to wait if it’s a quick thing to allow any function to raise an error, but otherwise I think we should get this in and potentially update the behavior in the future

Reply:
Thanks for looking into this @abidlabs  and for filing @pngwn !

Reply:
Yeh I would just get this in for now, we can tweak later when we fix that.